neo/Ledger/Blockchain.cs

```diff
@@ -0,0 +1,705 @@
+                stored_header_count += (uint)header_index.Count;
+                if (stored_header_count == 0)
+                {
+                    header_index.AddRange(store.GetBlocks().Find().OrderBy(p => p.Value.TrimmedBlock.Index).Select(p => p.Key));
+                }
+                else
+                {
+                    HashIndexState hashIndex = store.GetHeaderHashIndex().Get();
+                    if (hashIndex.Index >= stored_header_count)
+                    {
+                        DataCache<UInt256, BlockState> cache = store.GetBlocks();
+                        for (UInt256 hash = hashIndex.Hash; hash != header_index[(int)stored_header_count - 1];)
+                        {
+                            header_index.Insert((int)stored_header_count, hash);
+                            hash = cache[hash].TrimmedBlock.PrevHash;
+                        }
+                    }
+                }
+                if (header_index.Count == 0)
+                    Persist(GenesisBlock);
+                else
+                    UpdateCurrentSnapshot();
+                singleton = this;
+            }
+        }
+
+        public bool ContainsBlock(UInt256 hash)
+        {
+            if (block_cache.ContainsKey(hash)) return true;
+            return Store.ContainsBlock(hash);
+        }
+
+        public bool ContainsTransaction(UInt256 hash)
+        {
+            if (mem_pool.ContainsKey(hash)) return true;
+            return Store.ContainsTransaction(hash);
+        }
+
+        private void Distribute(object message)
+        {
+            foreach (IActorRef subscriber in subscribers)
+                subscriber.Tell(message);
+        }
+
+        public Block GetBlock(UInt256 hash)
+        {
+            if (block_cache.TryGetValue(hash, out Block block))
+                return block;
+            return Store.GetBlock(hash);
+        }
+
+        public UInt256 GetBlockHash(uint index)
+        {
+            if (header_index.Count <= index) return null;
+            return header_index[(int)index];
+        }
+
+        public static UInt160 GetConsensusAddress(ECPoint[] validators)
+        {
+            return Contract.CreateMultiSigRedeemScript(validators.Length - (validators.Length - 1) / 3, validators).ToScriptHash();
+        }
+
+        public IEnumerable<Transaction> GetMemoryPool()
+        {
+            return mem_pool.Values;
+        }
+
+        public Snapshot GetSnapshot()
+        {
+            return Store.GetSnapshot();
+        }
+
+        public Transaction GetTransaction(UInt256 hash)
+        {
+            if (mem_pool.TryGetValue(hash, out Transaction transaction))
+                return transaction;
+            return Store.GetTransaction(hash);
+        }
+
+        private void OnImport(IEnumerable<Block> blocks)
+        {
+            foreach (Block block in blocks)
```

Comment:
Suppose someone wants to stop the server in the middle of this loop? Shouldn't this have some logic to break in that case?

Reply:
I suppose as a workaround for now the sender could send messages in smaller groups of blocks.

Reply:
When the main thread exits, the threads of these actors are automatically terminated.

Reply:
Hmm, if you are doing things on Runtime.Notify; how do you ensure consistency across termination and restart? By sending in pieces at least you can control not stopping the main thread until a bit after completing the last requested import. I went this route and it seems fine.

Reply:
Consistency can be ensured because the program could have been terminated abnormally.

Reply:
I don't think that ensures consistency; the software should be fixed if it terminates abnormally. I suppose you can expect software sitting on top listening to runtime notify messages to heal itself from improper terminations, but we should strive for the behavior that it waits till all processing after a block completes before terminating.